libcnb-cargo/src/lib.rs

```diff
@@ -0,0 +1,191 @@
+    cargo_profile: CargoProfile,
+    target_triple: impl AsRef<str>,
+) -> Result<PathBuf, BuildError> {
+    let cargo_metadata = MetadataCommand::new()
+        .manifest_path(project_path.as_ref().join("Cargo.toml"))
+        .exec()
+        .map_err(BuildError::MetadataError)?;
+
+    let buildpack_cargo_package = cargo_metadata
+        .root_package()
+        .ok_or(BuildError::CouldNotFindRootPackage)?;
+
+    let target = match buildpack_cargo_package.targets.as_slice() {
+        [] => Err(BuildError::NoTargetsFound),
+        [single_target] => Ok(single_target),
+        _ => Err(BuildError::MultipleTargetsFound),
+    }?;
+
+    let mut cargo_args = vec!["build", "--target", target_triple.as_ref()];
+    match cargo_profile {
+        CargoProfile::Dev => {}
+        CargoProfile::Release => cargo_args.push("--release"),
+    }
+
+    let exit_status = Command::new("cargo")
+        .args(cargo_args)
+        .envs(
+            cross_compile::cross_compile_env(target_triple.as_ref())
+                .map_err(BuildError::CrossCompileError)?,
+        )
+        .current_dir(&project_path)
+        .spawn()
+        .and_then(|mut child| child.wait())
+        .map_err(BuildError::IoError)?;
+
+    if exit_status.success() {
+        Ok(cargo_metadata
+            .target_directory
+            .join(target_triple.as_ref())
+            .join(match cargo_profile {
+                CargoProfile::Dev => "debug",
+                CargoProfile::Release => "release",
+            })
+            .join(&target.name)
+            .into_std_path_buf())
+    } else {
+        Err(BuildError::UnexpectedExitStatus(exit_status))
+    }
+}
+
+#[derive(Debug)]
+pub enum BuildError {
+    IoError(std::io::Error),
+    UnexpectedExitStatus(ExitStatus),
+    CrossCompileError(CrossCompileError),
+    NoTargetsFound,
+    MultipleTargetsFound,
+    MetadataError(cargo_metadata::Error),
+    CouldNotFindRootPackage,
+}
+
+#[derive(Copy, Clone)]
+pub enum CargoProfile {
+    Dev,
+    Release,
+}
+
+/// Reads buildpack data from the given project path.
+pub fn read_buildpack_data(
+    project_path: impl AsRef<Path>,
+) -> Result<BuildpackData<Option<toml::Value>>, BuildpackDataError> {
+    let buildpack_toml_path = project_path.as_ref().join("buildpack.toml");
+
+    fs::read_to_string(&buildpack_toml_path)
+        .map_err(BuildpackDataError::IoError)
+        .and_then(|file_contents| {
+            toml::from_str(&file_contents).map_err(BuildpackDataError::DeserializationError)
+        })
+        .map(|buildpack_toml| BuildpackData {
+            buildpack_toml_path,
+            buildpack_toml,
+        })
+}
+
+#[derive(Debug)]
+pub enum BuildpackDataError {
+    IoError(std::io::Error),
+    DeserializationError(toml::de::Error),
+}
+
+pub struct BuildpackData<BM> {
+    pub buildpack_toml_path: PathBuf,
+    pub buildpack_toml: BuildpackToml<BM>,
+}
+
+/// Assembles a buildpack tarball and writes it to the given destination path.
```

Comment:
I'd like to discuss the approach of using a tarfile for output a bit more, however first some additional context may be useful (both for current and future readers), from the current status quo and also my experiments with packaging a couple of months ago.

At the moment there are quite a few issues in the way Cutlass works, which makes CI very slow (15 mins for just the maven specs in the JVM repo for example):
- In order to provide a streamlined local test UX, Cutlass builds and packages buildpacks as part of running tests, rather than it being a setup step prior to cutlass being invoked.
- Running Ruby tests in parallel is typically achieved via `parallel_split_test` or similar, which don't support a "before spawning processes" setup hook / "after all processes have exited" teardown hook.
- As such each concurrent test process has to perform the buildpack build/packaging, which at best duplicates work (even with caches) and at worst results in race conditions around setup/teardown.
- Due to those race conditions, concurrency is disabled in the JVM repo when using Cutlass.

Therefore, in whatever solution we have for packaging (this PR) and integration testing (eg future PR to add a Rust alternative to Cutlass), we need the ability to both perform setup/teardown once and only once, and still run tests concurrently. 

In addition, when testing buildpacks ideally we would just pass a directory path to `--buildpacks` (rather than an OCI image, as Cutlass does) since it reduces the setup/teardown complexity/time (and would likely avoid the race conditions in Cutlass), and actually makes `pack build` runs slightly faster (counter-intuitively, providing a packaged buildpack does not save pack any time).

However at present `pack build --buildpack` doesn't perform any of the steps that `pack buildpack package` performs, so effectively ignores `package.toml` and doesn't pull in dependent buildpacks of a meta-buildpack:
https://github.com/buildpacks/pack/issues/1320

As such if wanting to use `--buildpacks` with a meta-buildpack, one has to either:
1. Pass in all dependent buildpacks via additional `--buildpacks` arguments (though this breaks the defined `order`, so not viable for meta-buildpacks with multiple `order`s).
2. Package the meta-buildpack to a `.cnb` or OCI image first, using `pack buildpack package`, and then use that packaged buildpack in the `--buildpacks` argument instead.

ie: When testing we'll need to:
1. For buildpacks that aren't meta-buildpacks (and are written in Rust):
  - Cross-compile
  - Assemble the binary/symlinks/assets somewhere (eg a directory)
2. For meta-buildpacks:
  - Have a custom test `package.toml` that points to the development versions in the repo from (1)
  - `pack buildpack package` using that `package.toml`

At the moment this PR does (1) by creating a tarfile, rather than a directory.

I like the fact that assembling the tarball is portable (since it's using the `tar` crate, not shelling out to subprocesses), and reading the files straight off disk instead of performing intermediate copies.

However:
- For buildpacks that aren't meta-buildpacks, it seems creating a tarfile is more complex/slower than creating a directory (and unnecessary, since not affected by buildpacks/pack/issues/1320).
- For meta-buildpacks, we would have to use a different implementation anyway (that uses `pack buildpack package`), meaning we don't have any code reuse benefits.

Are there other benefits of using an output tarfile over a directory? I can imagine one might be for platforms where it's problematic to create a symlink (eg Windows), however there are other ways we could work around that, and I suspect it's not an issue in practice for people working on buildpacks written in Rust (since modern Windows does support symlinks in developer mode and also via WSL2 etc).

As a datapoint, using directory format via my bash script takes 450ms to no-op rebuild and package 5x buildpacks in a workspace, compared to 10s for `cargo libcnb package`. Of that 10s, 80% is spent in the tar file creation.

Reply:
Thanks for the writeup @edmorley, my experience is fairly similar. When doing this by hand, I almost never make a tarball since it's slow.

To that point, should we just not use `pack buildpack package -f` for everything vs making our own tarball?

Reply:
> To that point, should we just not use pack buildpack package -f for everything vs making our own tarball?

I'm not sure if I got this correctly, but I don't think adding another dependency on a command line tool is desirable, even if it's a tool that buildpack authors most likely have installed.

Reply:
So something to bear in mind for "create tarfile in Rust" vs `pack buildpack package -f` is that pack actually accepts two types of archives:
1. A single buildpack archive where the `bin/` directory is in the root of the archive.
2. A `.cnb` style archive where one or more buildpacks are at the `/cnb/buildpacks/<buildpack ID>/<buildpack version>/` locations described in the [distribution spec](https://github.com/buildpacks/spec/blob/main/distribution.md#buildpackage).

This PR implements style (1), which cannot be used for meta-buildpacks.

In order to support meta-buildpacks, style (2) will need to be used, which means that implementation starts to get more complex, and means we're potentially reimplementing parts of pack, if we stick with creating the tarfile in Rust. 

That may be something we still want to do, however it feels like a decision for later, when meta-buildpacks are actually supported by `cargo libcnb package`?

Reply:
Yeah, we have nothing for meta-buildpacks in libcnb as of now. As you said this is something for later.

Reply:
Buildpack assembly was replaced with a directory based approach in 7f52227. It currently does not any deletion of existing directories and will fail if the directory already exists. Happy to hear what people think is a good strategy for handling deletion when packaging in run multiple times.

Deleting the output directory is dangerous, if the user accidentally points the output path at `$HOME` we might delete their home directory. But forcing users to delete older versions by hand is also annoying.

Reply:
> [cutlass] eg creating OCI images, rather than leaving as directories 

- Fixed in https://github.com/heroku/cutlass/pull/37 (needs review please 🙏)
- Possible proposal to address the concurrency issue https://github.com/heroku/cutlass/pull/38

Reply:
> I'm not sure if I got this correctly, but I don't think adding another dependency on a command line tool is desirable, even if it's a tool that buildpack authors most likely have installed.

I think we should just get out of the `tar` business which you already did. If a user wants to create a tarball, they can use `pack` like everyone else vs us re-inventing the wheel.

Reply:
> Deleting the output directory is dangerous, if the user accidentally points the output path at $HOME we might delete their home directory. But forcing users to delete older versions by hand is also annoying.

We don't default to `$HOME`, I'm not sure I'm too worried about this edge case. I personally would prefer to have it clear / rebuild. We can add configure options if you really want.

Reply:
I'm marking this as resolved as the tar generation has been replaced by a directory based approach in 7f52227. Feel free to add review comments to the new code.